File: src/ollie/tracing/otel_setup.py
"""Configure TracerProvider + OllieSpanExporter."""

from __future__ import annotations

from typing import Any

_provider: Any = None
_processor: Any = None
_exporter: Any = None


def setup_tracer_provider(*, capture_content: bool = True) -> Any:
    """Install a TracerProvider with SimpleSpanProcessor → OllieSpanExporter."""
    global _provider, _processor, _exporter
    from opentelemetry import trace
    from opentelemetry.sdk.resources import Resource
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import SimpleSpanProcessor

    from ollie.tracing.exporter import OllieSpanExporter

    if _exporter is not None:
        _exporter.capture_content = capture_content
        _exporter.enabled = True
    else:
        _exporter = OllieSpanExporter(capture_content=capture_content)

    if _provider is not None:
        return _provider

    existing = trace.get_tracer_provider()
    if isinstance(existing, TracerProvider) and type(existing).__name__ == "TracerProvider":
        # Reuse process-global SDK provider (OTel disallows override).
        _processor = SimpleSpanProcessor(_exporter)
        existing.add_span_processor(_processor)
        _provider = existing
        return _provider

    resource = Resource.create({"service.name": "ollie-sdk"})
    provider = TracerProvider(resource=resource)
    _processor = SimpleSpanProcessor(_exporter)
    provider.add_span_processor(_processor)
    trace.set_tracer_provider(provider)
    _provider = provider
    return provider


def shutdown_tracer_provider() -> None:
    """Detach Ollie exporter; keep provider (OTel cannot replace it)."""
    global _processor, _exporter
    # Do not null _provider — OTel forbids replacing TracerProvider.
    # Exporter becomes a no-op when auto_instrument is off via instrumentor uninstall.
    if _exporter is not None:
        _exporter.enabled = False
    # Leave provider/processor in place; instrumentors control whether spans are created.

File: packages/ts/src/tracing/otelSetup.ts
import { trace } from "@opentelemetry/api";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { SimpleSpanProcessor, TracerProvider } from "@opentelemetry/sdk-trace";

import { OllieSpanExporter } from "./exporter.js";

let provider: TracerProvider | null = null;
let exporter: OllieSpanExporter | null = null;

export async function setupTracerProvider(options?: { captureContent?: boolean }): Promise<TracerProvider> {
  if (exporter) {
    exporter.captureContent = options?.captureContent !== false;
    exporter.enabled = true;
  } else {
    exporter = new OllieSpanExporter({ captureContent: options?.captureContent });
  }

  if (provider) return provider;

  provider = new TracerProvider({
    resource: resourceFromAttributes({ "service.name": "ollie-sdk-ts" }),
    spanProcessors: [new SimpleSpanProcessor({ exporter })],
  });
  trace.setGlobalTracerProvider(provider);
  return provider;
}

export async function shutdownTracerProvider(): Promise<void> {
  if (exporter) exporter.enabled = false;
  if (provider) {
    await provider.shutdown();
    provider = null;
  }
}
